lib/matobjplist.gi

```diff
@@ -1028,132 +783,41 @@ InstallMethod( ZeroMutable, "for a plist matrix",
-        #Error("OneSameMutability: Matrix is not square");
-        #return;
-        return fail;
-    fi;
-    o := IdentityMatrix(m![RLPOS],m);
-    if not IsMutable(m) then
-        MakeImmutable(o);
-    fi;
-    return o;
+    return IsOne(m![ROWSPOS]); # TODO: check that this results in "optimal" code
   end );
 
 InstallMethod( OneMutable, "for a plist matrix",
   [ IsPlistMatrixRep ],
   function( m )
-    if m![RLPOS] <> Length(m![ROWSPOS]) then
+    if NrCols(m) <> NrRows(m) then
         #Error("OneMutable: Matrix is not square");
         #return;
         return fail;
     fi;
-    return IdentityMatrix(m![RLPOS],m);
-  end );
-
-InstallMethod( OneImmutable, "for a plist matrix",
-  [ IsPlistMatrixRep ],
-  function( m )
-    local o;
-    if m![RLPOS] <> Length(m![ROWSPOS]) then
-        #Error("OneImmutable: Matrix is not square");
-        #return;
-        return fail;
-    fi;
-    o := IdentityMatrix(m![RLPOS],m);
-    MakeImmutable(o);
-    return o;
+    return NewIdentityMatrix(IsPlistMatrixRep,NrCols(m));
   end );
 
-# For the moment we delegate to the fast kernel arithmetic for plain
-# lists of plain lists:
-
 InstallMethod( InverseMutable, "for a plist matrix",
   [ IsPlistMatrixRep ],
   function( m )
     local n;
-    if m![RLPOS] <> Length(m![ROWSPOS]) then
-        #Error("InverseMutable: Matrix is not square");
-        #return;
-        return fail;
-    fi;
-    # Make a plain list of lists:
-    n := List(m![ROWSPOS],x->x![ELSPOS]);
-    n := InverseMutable(n);  # Invert!
-    if n = fail then return fail; fi;
-    return Matrix(n,Length(n),m);
-  end );
-
-InstallMethod( InverseImmutable, "for a plist matrix",
-  [ IsPlistMatrixRep ],
-  function( m )
-    local n;
-    if m![RLPOS] <> Length(m![ROWSPOS]) then
-        #Error("InverseMutable: Matrix is not square");
-        #return;
-        return fail;
-    fi;
-    # Make a plain list of lists:
-    n := List(m![ROWSPOS],x->x![ELSPOS]);
-    n := InverseMutable(n);  # Invert!
-    if n = fail then return fail; fi;
-    n := Matrix(n,Length(n),m);
-    MakeImmutable(n);
-    return n;
-  end );
-
-InstallMethod( InverseSameMutability, "for a plist matrix",
-  [ IsPlistMatrixRep ],
-  function( m )
-    local n;
-    if m![RLPOS] <> Length(m![ROWSPOS]) then
-        #Error("InverseMutable: Matrix is not square");
-        #return;
+    if NrCols(m) <> NrRows(m) then
         return fail;
     fi;
     # Make a plain list of lists:
-    n := List(m![ROWSPOS],x->x![ELSPOS]);
-    n := InverseMutable(n);  # Invert!
+    n := InverseMutable(m![ROWSPOS]);
     if n = fail then return fail; fi;
-    n := Matrix(n,Length(n),m);
-    if not IsMutable(m) then
-        MakeImmutable(n);
-    fi;
-    return n;
+    # FIXME: if the base domain is not a field, e.g. Integers,
+    # then the inverse we just computed may not be defined over
+    # that base domain. Should we detect this here? Or how else
+    # will we deal with this in general?
```

Comment:
I think this is first of all a question about what the operation `NewMatrix` promises. Perhaps we need `NewMatrix` and `NewMatrixNC`? (And then the question is whether we call `NewMatrix` or `NewMatrixNC` here.)